db/bigtable_eth1.go

```diff
@@ -4001,7 +4014,7 @@ func (bigtable *Bigtable) GetBlockKeys(blockNumber uint64, blockHash []byte) ([]
 	}
 
 	if row == nil {
-		return nil, fmt.Errorf("keys for block %v not found", blockNumber)
+		return make([]string, 0), nil
```

Comment:
**question/issue:** What was your intended behaviour here?

Before we threw an error if the block which was supposed to be deleted could not be found (i.e. no rows) but now we don't.

**Side note:** I see that the result is then immediately used to check if the length is zero.
For that a nil slice or an empty slice makes no difference, both return 0.
Nil slice and empty slice are only different for json encoding.

Was the reason to remove the error so that there is no confusion because you already deleted the contract states info from the bigtable first?
If that is the case why not delete the normal info that was already in `DeleteBlock` first and then delete the contract states?
In that case if `GetBlockKeys` throws this error nothing gets deleted for the block.

Reply:
The motivation for this change was to be able to handle empty blocks correctly. So `the block which was supposed to be deleted could not be found (i.e. no rows)` is not really always true I think, and this was my attempt of dealing with it. While not relevant for the main networks, it still is for the local testnet and potentially other forks with smaller communities.

Note that the block itself does get deleted anyways (see below though :sweat_smile:), this is only about [keys that were written by this block (entities & indices](https://github.com/gobitfly/eth2-beaconchain-explorer/pull/2618/files#diff-143a07e5aa31442d87e5dd9584d1525fc98e8b2fef161a08af276d65f60184acR4076), of which there might be none.

But after looking at it again, there's still an issue here: Because of the zero-length check early-out in `DeleteBlock`, we actually never [delete the block](https://github.com/gobitfly/eth2-beaconchain-explorer/pull/2618/files#diff-143a07e5aa31442d87e5dd9584d1525fc98e8b2fef161a08af276d65f60184acR4098). So I'll remove that early-out.
<br />
Just a side note (as the premise isn't true), but just to clarify:
> why not delete the normal info that was already in DeleteBlock first and then delete the contract states?

Because I think removing the block from `tableBlocks` feels like the finishing move of that function, confirming its successful deletion. If anything else before that has failed, that step should not be executed but rather there needs to be some investigation what's wrong.
<br />
I could also be missing something of course; Lmk if you agree and this answers your question, or if there's anything else to be taken care of :)

Reply:
You mean empty blocks like this?
https://beaconcha.in/block/15596410#overview

Don't we still save block keys even for such empty blocks?
We save them here
https://github.com/gobitfly/eth2-beaconchain-explorer/blob/3744839a466c64bfe48ef4dbc9f0e4b50f8cc9a3/db/bigtable_eth1.go#L694
and get data from all `transform` methods
https://github.com/gobitfly/eth2-beaconchain-explorer/blob/3744839a466c64bfe48ef4dbc9f0e4b50f8cc9a3/db/bigtable_eth1.go#L679
The list of them is here
https://github.com/gobitfly/eth2-beaconchain-explorer/blob/a076907a039c3a3e9073a222a4a2128f0833262c/cmd/eth1indexer/main.go#L173
which includes
https://github.com/gobitfly/eth2-beaconchain-explorer/blob/3744839a466c64bfe48ef4dbc9f0e4b50f8cc9a3/db/bigtable_eth1.go#L761
for generic blocks information like its hash.

Reply:
I reverted this change; As discussed, I misinterpreted the consequences of another issue related to general reorg handling.